{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TME 8 : Mini-projet Detection de motifs\n",
    "\n",
    "\n",
    "## Localisation de pattern (motifs) dans le genome\n",
    "\n",
    "Nous allons implémenter plusieurs algorithmes afin de localiser un motif d'intérêt dans un génome."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### A -  Boyer Moore Algorithm \n",
    "1\\. Nous allons réutiliser les fonctions du TME6 pour générer une séquence artificiel de taille ``n``, et implanter un motif de taille ``k`` dans une position aléatoire de la séquence. Vous pouvez aussi faire varier ``v`` positions pour permettre certaines mutations\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "import numpy as np\n",
    "import time\n",
    "\n",
    "nuc = ('A', 'C', 'G', 'T')\n",
    "\n",
    "t = 1\n",
    "n = 100 #longuer des sequence\n",
    "k = 8 #longuer du motif\n",
    "v = 0 \n",
    "\n",
    "def generateRandomSequences(n, t):\n",
    "    sequences = []\n",
    "    i = 0\n",
    "    j = 0\n",
    "    while i < t:\n",
    "        seq = \"\"\n",
    "        while j < n:\n",
    "            seq = seq+nuc[random.randint(0,3)]\n",
    "            j += 1\n",
    "        sequences.append(seq)\n",
    "        i += 1\n",
    "        j = 0\n",
    "   \n",
    "    return sequences\n",
    "\n",
    "# RENVOIE UNE LISTE DE LISTE DE MOTIFS...\n",
    "def generateMotifs(k, nM):\n",
    "    return [generateRandomSequences(k, True) for rs in range(nM)]\n",
    "\n",
    "def gen_n_substitMotifs(v, k, t):\n",
    "    motifs = generateMotifs(k, 1)\n",
    "    print(motifs[0])\n",
    "    motif = ''.join(motifs[0])\n",
    "    motifs = []\n",
    "    i=0\n",
    "    while i < t:\n",
    "        j = 0\n",
    "        motif_a_modif = list(motif.strip())\n",
    "        while j < v:\n",
    "            motif_a_modif[random.randint(0, k - 1)] = nuc[random.randint(0,3)]\n",
    "            motif_add = ''.join(motif_a_modif)\n",
    "            j+=1\n",
    "        motifs.append(motif_add)\n",
    "        i+=1\n",
    "    return motifs\n",
    "    \n",
    "def implantMotifVar(k, v, t, n):\n",
    "    #sequences aleatoires\n",
    "    sequences = generateRandomSequences(n, t)\n",
    "    if v != 0:\n",
    "        motifs = gen_n_substitMotifs(v, k, t)\n",
    "    else:\n",
    "        motifs_generated = generateMotifs(k, 1)\n",
    "        # il faut remettre dans une simple liste vu QUE GENERATE_MOTIFS fait une liste de LISTE\n",
    "        motifs = motifs_generated[0]\n",
    "    modified_sequences = []\n",
    "    i = 0\n",
    "    while i < t:\n",
    "        seq = sequences[i]\n",
    "        if v != 0:\n",
    "            motif = motifs[i]\n",
    "        else:\n",
    "            motif = motifs[0]\n",
    "        rand_pos = random.randint(0,len(seq))\n",
    "        seq = seq[0:rand_pos] + motif + seq[rand_pos+1:len(seq)]\n",
    "        modified_sequences.append(seq)\n",
    "        i+=1\n",
    "    return motifs, modified_sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "motif : TGGCCTAC\nsequence : CTCCCCTGGCCTACGTGTAGTGGCCGCCGATATCTCACACATCTAAGTCACATTTGGTCGGTGCGCACTTGTACCAAAAAAATTCGAGCCCCTTGCCCACAGGTCCG\n"
     ]
    }
   ],
   "source": [
    "#generate and implant variavel motifs\n",
    "motifs, sequences = implantMotifVar(k, v, t, n)\n",
    "motif = motifs[0]\n",
    "sequence = sequences[0]\n",
    "print(\"motif : {}\".format(motif))\n",
    "print(\"sequence : {}\".format(sequence))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2\\. Boyer Moore est une combinaison de deux approches : bad character et good Suffix. Nous allons commencer par la première approche. Définissez une fonction ``badCharacter`` pour construire la badTable d'un motif. Vous pouvez l'implémenter comme dictionnaire, voir un exemple ci-dessous.\n",
    "\n",
    "<img src=\"boyer-moore-badTable.png\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "{'T': 2, 'G': 5, 'C': 1, 'A': 1}"
      ]
     },
     "metadata": {},
     "execution_count": 24
    }
   ],
   "source": [
    "def badCharacter(motif):\n",
    "    '''\n",
    "    The preprocessing function for Boyer Moore's bad character heuristic\n",
    "    '''\n",
    "    badChar = {}\n",
    "    i = 0\n",
    "    while i < len(motif):\n",
    "        char = motif[i]\n",
    "        badChar[char] = max(1, len(motif) - i - 1)\n",
    "        i+=1\n",
    "    \n",
    "    #print(badChar)\n",
    "    return badChar\n",
    "           \n",
    "badCharacter(\"TATGTG\") #{'T': 1, 'A': 4, 'G': 1}\n",
    "badCharacter(motif)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3\\. Développer la version de l'algorithme Boyer Moore qui utilise \"bad character\" la règle de caractère incorrect."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "-1\n4\n3\n-1\n6\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "def search(genome, motif):\n",
    "    '''\n",
    "    A pattern searching function that uses Bad Character Heuristic of Boyer Moore Algorithm\n",
    "    '''\n",
    "    bc = badCharacter(motif)\n",
    "    pos = 0\n",
    "    i = 0\n",
    "    cpt = 0\n",
    "    while pos + (len(motif) - i) < len(genome):\n",
    "        cpt += 1\n",
    "        char_gen = genome[pos + (len(motif) - i) - 1]\n",
    "        char_mot = motif[len(motif) - i - 1]\n",
    "        if char_gen != char_mot:\n",
    "            if char_gen not in bc:\n",
    "                pos += len(motif)\n",
    "            else:\n",
    "                pos += bc[char_gen]\n",
    "            i = 0\n",
    "        else:\n",
    "            i += 1\n",
    "        if i >= len(motif):\n",
    "            break\n",
    "    \n",
    "    #print(\"Nombre de tours de boucle : \",cpt)\n",
    "    #print(pos)\n",
    "    if pos + (len(motif) - i) >= len(genome):\n",
    "        return -1\n",
    "\n",
    "    return pos\n",
    "\n",
    "print(search(\"ATCGCA\", \"TA\")) #-1\n",
    "            \n",
    "print(search(\"ATGTCCGTC\", \"CCG\")) #4\n",
    "\n",
    "print(search(\"ATCGCA\", \"GC\")) #3\n",
    "\n",
    "print(search(\"CGTGGCCAAGGAGCT\", \"GTGGCCAA\"))\n",
    "\n",
    "print(search(sequence, motif))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4\\. Testez la version de l'exercice 3 sur les données artificielles généré dans 1, combien de comparaison inutile avez-vous évité avec cette version? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "TGGCCTAC\nCTCCCCTGGCCTACGTGTAGTGGCCGCCGATATCTCACACATCTAAGTCACATTTGGTCGGTGCGCACTTGTACCAAAAAAATTCGAGCCCCTTGCCCACAGGTCCG\n6\nTGGCCTAC\n"
     ]
    },
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "6"
      ]
     },
     "metadata": {},
     "execution_count": 26
    }
   ],
   "source": [
    "print(motif)\n",
    "print(sequence)\n",
    "pos = search(sequence, motif)\n",
    "print(pos)\n",
    "print(sequence[pos:pos + k])\n",
    "sequence.find(motif)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reponse: On obtient le bon motif. Nous aurions fait 92 comparaisons normalement, (100(taille sequence) - 8(taille motif)), ici nous en avons fait 24 nous en avons donc eviter 68."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "5\\. Nous allons implémenter la règle de good suffixe. Pour cela, nous allons pré-processer le motif afin de trouver tous les suffixe de taille k que sont aussi de préfixe. Faire la fonction goodSuffix qui étant donnée un motif renvoie un dictionnaire ou les clés sont de suffixes et les valeurs sont la distance au préfixe le plus proche, voir un exemple ci-dessous. d2 est la distance entre le suffixe de taille k et le son occurence la plus proche."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"suffix.png\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "{'B': 2, 'AB': 4, 'BAB': 4, 'CBAB': 4, 'BCBAB': 4}\n{'B': 2, 'AB': 5, 'BAB': 5, 'OBAB': 5, 'AOBAB': 5}\n"
     ]
    }
   ],
   "source": [
    "\n",
    "def goodSuffix(motif):\n",
    "    gs={}\n",
    "    i=0\n",
    "    suffixe=\"\"\n",
    "    tmplength=0\n",
    "    while i<len(motif)-1:\n",
    "        suffixe=motif[len(motif)-i-1]+suffixe\n",
    "        if i!=len(motif)-1:\n",
    "            char_1=motif[len(motif)-i-2]\n",
    "        j=0\n",
    "        \"\"\"\n",
    "        on essaie de trouver le suffixe dans le reste du motif\n",
    "        \"\"\"\n",
    "        while j+i<len(motif)-1:\n",
    "            if suffixe==motif[j:j+i+1]:\n",
    "                if j!=0 and i!=len(motif)-1:\n",
    "                    if char_1!=motif[j-1]:\n",
    "                        gs[suffixe]=len(motif)-i-j-1\n",
    "                else :\n",
    "                    gs[suffixe]=len(motif)-i-j-1\n",
    "            j+=1\n",
    "        i+=1\n",
    "        \n",
    "        \"\"\"\n",
    "        cas échéant : on essaie de voir si la fin du suffixe matche avec le préfixe, sinon on lui associe la longueur du motif\n",
    "        \"\"\"\n",
    "        maxl=0\n",
    "        if suffixe not in gs:\n",
    "            for s in gs:\n",
    "                if s == motif[0:len(s)] and maxl<len(s):\n",
    "                    maxl=len(s)\n",
    "            if maxl!=0:\n",
    "                gs[suffixe]=len(motif)-maxl\n",
    "            else:\n",
    "                gs[suffixe]=len(motif)\n",
    "    return gs\n",
    "\n",
    "gs = goodSuffix(\"ABCBAB\"); print (gs)  #{'B': 2, 'AB': 4, 'BAB': 4, 'CBAB': 4, 'BCBAB': 4}\n",
    "gs = goodSuffix(\"BAOBAB\"); print (gs)  #{'B': 2, 'AB': 2, 'BAB': 5, 'OBAB': 5, 'AOBAB': 5}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "6\\. Développer la version de l'algorithme Boyer Moore qui utilise les deux règles \"bad character\" et  \"good suffixe\". Avez-vous amélioré votre algorithme? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "-1\n6\n0.0005009174346923828\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "def search(genome, motif):\n",
    "    '''\n",
    "    A pattern searching function that uses Bad Character\n",
    "    Heuristic of Boyer Moore Algorithm\n",
    "    '''\n",
    "    start = time.time()\n",
    "    bc = badCharacter(motif) \n",
    "    gs = goodSuffix(motif)\n",
    "    \n",
    "    res = -1\n",
    "    cpt = 0\n",
    "    pos = 0\n",
    "    offset = 1\n",
    "    while pos < len(genome) - len(motif):\n",
    "        cpt += 1\n",
    "        char_genome = genome[pos + len(motif) - offset]\n",
    "        char_motif = motif[len(motif) - offset]\n",
    "        if char_genome != char_motif:\n",
    "            suffixe_genome = genome[pos + len(motif) - offset:pos + len(motif)]\n",
    "            badchar = bc[char_genome]\n",
    "            if suffixe_genome in gs:\n",
    "                goodsuff = gs[suffixe_genome]\n",
    "            else:\n",
    "                goodsuff = -1\n",
    "            pos += max(badchar, goodsuff)\n",
    "            offset = 1\n",
    "        else:\n",
    "            offset += 1\n",
    "        if offset > len(motif):\n",
    "            res = pos\n",
    "            break\n",
    "\n",
    "    \n",
    "    return res\n",
    "              \n",
    "#search(\"ACAAABABABC\", \"ABABAB\")\n",
    "print(search(\"GGACGGCGTGGCCAAGGAGCT\", \"GTGGCCAA\"))\n",
    "start = time.time()\n",
    "print(search(sequence, motif))\n",
    "end = time.time()\n",
    "print(end - start)\n",
    "       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Reponse : Sur cette sequence l''algoritme ne montre pas d''amelioration"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "7\\. Tester l'algorithme **Boyer Moore Algorithm** sur vos données de chipSeq. Vous pouvez chercher de motifs trouvés par les algorithms développé précédemment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "def readFasta(genome):\n",
    "    sequence = []\n",
    "    file = open(genome, \"r\")\n",
    "    sequences = []\n",
    "    seq = \"\"\n",
    "    for s in file:\n",
    "        if s[0] != \">\":\n",
    "            seq += s.strip().upper()\n",
    "        else:\n",
    "            sequences.append(seq)\n",
    "            seq = \"\"\n",
    "    return sequences[1:]\n",
    "genome = \"Sequence_by_Peaks_6.fasta\"\n",
    "sequences = readFasta(genome)\n",
    "#print(sequences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "0.011501789093017578\n[-1, 180, -1, -1, -1, -1, -1, -1, 370, -1, -1, -1, 489, 197, -1, -1, -1, -1, -1, 155, 507, -1, -1, 373, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]\n"
     ]
    }
   ],
   "source": [
    "start = time.time()\n",
    "liste_res = []\n",
    "for seq in sequences:\n",
    "    liste_res.append(search(seq, \"TTTGCTCA\"))\n",
    "    #print(\"TTTGCTCA :   \",search(seq, \"TTTGCTCA\"))\n",
    "end = time.time()\n",
    "print(end - start)\n",
    "print(liste_res)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### B -  Index of fixed length words\n",
    "8\\. Faire une fonction pour indexer les positions d'occurrences de tous les mots de taille k dans un texte, la fonction doit renvoyer un dictionnaire ou les clés sont les mots et les valeurs les positions dans le texte."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "{'ACT': [0, 3, 6, 9], 'CTA': [1, 4, 7], 'TAC': [2, 5, 8]}"
      ]
     },
     "metadata": {},
     "execution_count": 56
    }
   ],
   "source": [
    "def indexTable(k, texte):\n",
    "    indexes  = {}\n",
    "\n",
    "    i = 0\n",
    "    while i < len(texte) - k + 1:\n",
    "        mot = texte[i:i + k]\n",
    "        if mot not in indexes:\n",
    "            indexes[mot] = [i]\n",
    "        else:\n",
    "            indexes[mot].append(i)\n",
    "        i += 1\n",
    "            \n",
    "    return indexes\n",
    "\n",
    "indexTable(3, \"ACTACTACTACT\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "9\\. Faire une fonction pour découper un motif en sous-parties de taille k"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "['ATC', 'TCA', 'CAT', 'ATC', 'TCA', 'CAG', 'AGT']"
      ]
     },
     "metadata": {},
     "execution_count": 57
    }
   ],
   "source": [
    "def seeds(k, motif):\n",
    "    sous_motifs = []\n",
    "    if k > len(motif):\n",
    "        print(\"Erreur seeds : K est plus grand que le motif\")\n",
    "        return -1\n",
    "    i = 0\n",
    "    while i < len(motif) - k + 1:\n",
    "        sous_motifs.append(motif[i:i+k])\n",
    "        i += 1\n",
    "    return sous_motifs\n",
    "\n",
    "\n",
    "seeds(3, \"ATCATCAGT\")\n",
    "   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "10\\. Faire une fonction pour chercher un motif dans un texte utilisant la table des indexes de la question 7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "0.0\n6\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "\n",
    "def searchIndexTable(texte, k, motif):\n",
    "    #start = time.time()\n",
    "    sous_motifs = seeds(k, motif)\n",
    "    index_table = indexTable(k, texte)\n",
    "    cpt = 0\n",
    "    res = -1\n",
    "    i = 0\n",
    "    while i < len(sous_motifs):\n",
    "        sousM = sous_motifs[i]\n",
    "        if sousM in index_table:\n",
    "            cpt += 1\n",
    "            indexes = index_table[sousM]\n",
    "            for index in indexes:\n",
    "                genome = texte[index - i: index + k]\n",
    "                if genome == motif:\n",
    "                    res = index - i\n",
    "                    break\n",
    "        if res != -1:\n",
    "            break       \n",
    "        i += 1\n",
    "    #print(\"cpt :\",cpt)\n",
    "    #end = time.time()\n",
    "    #print(end - start)\n",
    "    return res\n",
    "\n",
    "k = 2 \n",
    "start = time.time()\n",
    "res = searchIndexTable(sequence, k, motif)\n",
    "end=time.time()\n",
    "print(end - start)\n",
    "print(res)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "11\\. Tester l'algorithme **Index of fixed length** words sur vos données de chipSeq. Puis générér le LOGO du motif trouvé"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "TTTGCTCA :    -1\nTTTGCTCA :    180\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    370\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    489\nTTTGCTCA :    197\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    155\nTTTGCTCA :    507\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    373\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\nTTTGCTCA :    -1\n"
     ]
    }
   ],
   "source": [
    "liste_res = []\n",
    "start = time.time()\n",
    "for seq in sequences:\n",
    "    liste_res.append(searchIndexTable(seq, 2, \"TTTGCTCA\"))\n",
    "    #print(\"TTTGCTCA :   \",searchIndexTable(seq, 2, \"TTTGCTCA\"))\n",
    "end = time.time()\n",
    "print(end - start)\n",
    "print(lis)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### C -  Matrice de fréquences\n",
    "12\\.Nous allons implémenter la recherche de motif par matrice de fréquence. Nous allons utiliser les matrices déjà fabriqué que vous pouvez telecharcher du site http://jaspar.genereg.net/\n",
    "Une fois que vous avez chargé votre matrice de fréquence, vous devez la transformer en matrice de probabilité"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "[[0.462, 0.024, 0.0, 1.0, 0.0, 0.007, 0.09, 0.6], [0.086, 0.016, 0.022, 0.0, 1.0, 0.0, 0.198, 0.138], [0.387, 0.0, 0.803, 0.0, 0.0, 0.993, 0.0, 0.162], [0.065, 0.96, 0.175, 0.0, 0.0, 0.0, 0.712, 0.1]]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "nuc = ('A', 'C', 'G', 'T')\n",
    "k = 8\n",
    "\n",
    "matrice = np.array([[462.00,24.00,0.00,1000.00,0.00,7.00,90.00,600.00],\n",
    "                    [86.00,16.00,22.00,0.00,1000.00,0.00,198.00,138.00],\n",
    "                    [387.00,0.00,803.00,0.00,0.00,993.00,0.00,162.00],\n",
    "                    [65.00,960.00,175.00,0.00,0.00,0.00,712.00,100.00]])\n",
    "\n",
    "#Computing wi(a)\n",
    "def pwm(P, k):\n",
    "    res=[]\n",
    "    i=0\n",
    "    j=0\n",
    "    while i<4:  #Il y aura toujours que 4 bases (A,C,G,T) \n",
    "        tempres=[]\n",
    "        while j<k: #k etant la longueur des motifs, c'est aussi la longueur des profils\n",
    "            tempres.append(P[i][j]/(P[0][j]+P[1][j]+P[2][j]+P[3][j]))\n",
    "            j+=1\n",
    "        j=0\n",
    "        i+=1\n",
    "        res.append(tempres)\n",
    "    return res\n",
    "\n",
    "\n",
    "PWM = pwm(matrice, k); print (PWM)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "12\\. Déterminer les paramètres f(0)(b) du modèle nul, où \n",
    "\\begin{equation}\n",
    "f^{(0)}(b) = \\frac 1L \\sum_{i=0}^{L-1} \\omega_i(b)\\ ,\n",
    "\\end{equation}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "{'A': 0.272875, 'C': 0.1825, 'G': 0.29312499999999997, 'T': 0.2515}\n"
     ]
    }
   ],
   "source": [
    "def f0_calcule(PWM, k):\n",
    "\tres = {}\n",
    "\ti = 0\n",
    "\tfor ligne in PWM:\n",
    "\t\tsomme_ligne = 0\n",
    "\t\tfor colonne in ligne:\n",
    "\t\t\tsomme_ligne += colonne\n",
    "\t\tres[nuc[i]] = somme_ligne / k\n",
    "\t\ti+=1\n",
    "\t\n",
    "\treturn res\n",
    "\n",
    "f_0 = f0_calcule(PWM, k)\n",
    "print (f_0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "13\\.Faites une fonction pour calculer log-rapport de vraisemblancee, d'une sequence de taille k.\n",
    "\\begin{equation}\n",
    "\\label{eq:ll}\n",
    "\\ell(b_0,...,b_{k-1}) = \\log_2 \\frac {P(b_0,...,b_{k-1} | \\omega )\n",
    "}{P^{(0)}(b_0,...,b_{k-1})}\n",
    "= \\sum_{i=0}^{k-1} \\log_2 \\frac {\\omega_i(b_i)}{f^{(0)}(b_i)}\\ .\n",
    "\\end{equation}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [],
   "source": [
    "#PFM de Mac1, associé au peaks6\n",
    "Mac1 = np.array([[108.00, 89.00, 0.00, 115.00, 0.00, 0.00, 0.00, 958.00],\n",
    "                 [0.00, 0.00, 0.00, 0.00, 1714.00, 0.00, 1714.00, 0.00],\n",
    "                 [0.00, 219.00, 0.00, 1599.00, 0.00, 0.00, 0.00, 655.00],\n",
    "                 [1632.00, 1451.00, 1714.00, 0.00, 0.00, 1714.00, 0.00, 113.00]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {},
   "outputs": [],
   "source": [
    "import math\n",
    "\n",
    "def loglikehood(sequences, PWM, f_0):\n",
    "    #print(k)\n",
    "    res = []\n",
    "    #nb_motifs = 0\n",
    "    for seq in sequences:\n",
    "        liste_motifs = seeds(k, seq)\n",
    "        #nb_motifs += len(liste_motifs)\n",
    "        for motif in liste_motifs:\n",
    "            j = 0\n",
    "            score = 0\n",
    "            while j < len(motif):\n",
    "                lettre = motif[j]\n",
    "                prob_lettre = PWM[nuc.index(lettre)][j]\n",
    "                if prob_lettre == 0.0:\n",
    "                    #aucune probabilité d'obtenir cette lettre\n",
    "                    score = -1\n",
    "                    break\n",
    "                f0_lettre = f_0[lettre]\n",
    "                score += math.log2(prob_lettre / f0_lettre)\n",
    "                j += 1\n",
    "            if score > 0 and motif not in res:\n",
    "                res.append(motif)\n",
    "    #print(nb_motifs)\n",
    "    return res"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "14\\.Tester l'algorithme sur vos données de chipSeq. Vous devez parcourir les séquences et extraire les motifs ayant  log-rapport de vraisemblance positive "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import Counter\n",
    "m = 5\n",
    "\n",
    "def removeLowComplexe(motifs, minrep):\n",
    "    validMotif = []\n",
    "    for motif in motifs:\n",
    "        frequence = Counter(motif)\n",
    "        # temoin pour savoir si une lettre est repetee ou non\n",
    "        temoin = True\n",
    "        for lettre, value in frequence.items():\n",
    "            if value >= minrep:\n",
    "                temoin = False\n",
    "        if temoin == True:\n",
    "            validMotif.append(motif)\n",
    "                \n",
    "    return validMotif\n",
    "\n",
    "def removeLowComplexePair(motifs):\n",
    "    validMotif = []\n",
    "    for motif in motifs:\n",
    "       frequence = Counter(motif)\n",
    "       if len(frequence) > 2:\n",
    "           validMotif.append(motif)\n",
    "\n",
    "    return validMotif"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "14\n['AATACTCT', 'TTTGCTCA', 'ATTGCTCT', 'TGTACTCT', 'TTTGCTCG', 'ATTGCTCA', 'TATGCTCA', 'TGTGCTCA', 'ATTACTCA', 'AGTACTCA', 'TGTACTCG', 'AGTACTCT', 'AATGCTCT', 'ATTACTCT']\noui\n"
     ]
    }
   ],
   "source": [
    "\n",
    "motif_mac1 = \"TTTGCTCA\"\n",
    "PWM_mac = pwm(Mac1, k)\n",
    "f_0_mac = f0_calcule(PWM_mac, k)\n",
    "liste_res = loglikehood(sequences, PWM_mac, f_0_mac)\n",
    "#remove motifs non interessants\n",
    "liste_res = removeLowComplexe(liste_res, m)\n",
    "liste_res = removeLowComplexePair(liste_res)\n",
    "print(len(liste_res))\n",
    "print(liste_res)\n",
    "if motif_mac1 in liste_res:\n",
    "    print(\"oui\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### D -  Localisation de facteur de transcriptions\n",
    "15\\. Une fois que vous avez localisé le motif dans le donnée de chipSeq, nous allons identifier les fonctions de gènes qui sont associées aux facteurs de transcription. Pour cela, nous allons utiliser le fichier qui contient tout le génome (C_glabrata_1000bp_only.fasta), localiser le motif (vous pouvez utiliser un des algos précédent), et identifier les gènes. Ensuite nous pouvons vérifier leurs fonctions dans le fichier \"C_glabrata_annotations.csv\" à l'aide de ses identifiant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {
    "tags": [
     "outputPrepend"
    ]
   },
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "1347656\n",
      "0.004999876022338867\n",
      "0.0050013065338134766\n",
      "0.004500865936279297\n",
      "0.004500389099121094\n",
      "0.00550079345703125\n",
      "0.005001068115234375\n",
      "0.0050008296966552734\n",
      "0.0050008296966552734\n",
      "0.0050008296966552734\n",
      "0.0050008296966552734\n",
      "0.005000591278076172\n",
      "0.005000591278076172\n",
      "0.0050013065338134766\n",
      "0.004500865936279297\n",
      "0.004500389099121094\n",
      "0.004500389099121094\n",
      "0.046508073806762695\n",
      "0.005000591278076172\n",
      "0.005001068115234375\n",
      "0.005000114440917969\n",
      "0.004500389099121094\n",
      "0.004500627517700195\n",
      "0.004500865936279297\n",
      "0.004500627517700195\n",
      "0.0050008296966552734\n",
      "0.004500627517700195\n",
      "0.0050008296966552734\n",
      "0.005001068115234375\n",
      "0.005000591278076172\n",
      "0.004500865936279297\n",
      "0.0050008296966552734\n",
      "0.004500627517700195\n",
      "0.0049991607666015625\n",
      "0.0050008296966552734\n",
      "0.0050008296966552734\n",
      "0.004500865936279297\n",
      "0.0050008296966552734\n",
      "0.0045011043548583984\n",
      "0.004500627517700195\n",
      "0.0045011043548583984\n",
      "0.005501747131347656\n",
      "0.004999876022338867\n",
      "0.0050013065338134766\n",
      "0.006000995635986328\n",
      "0.0050008296966552734\n",
      "0.005501985549926758\n",
      "0.004999637603759766\n",
      "0.005012035369873047\n",
      "0.004499673843383789\n",
      "0.005000114440917969\n",
      "0.0045011043548583984\n",
      "0.0050008296966552734\n",
      "0.005000591278076172\n",
      "0.005001068115234375\n",
      "0.005000591278076172\n",
      "0.005000114440917969\n",
      "0.0050008296966552734\n",
      "0.005501747131347656\n",
      "0.005000114440917969\n",
      "0.004500865936279297\n",
      "0.004999876022338867\n",
      "0.04700827598571777\n",
      "0.005001068115234375\n",
      "0.005000591278076172\n",
      "0.0045011043548583984\n",
      "0.004500389099121094\n",
      "0.0050008296966552734\n",
      "0.0050008296966552734\n",
      "0.005001068115234375\n",
      "0.005000591278076172\n",
      "0.0050008296966552734\n",
      "0.005501270294189453\n",
      "0.0055010318756103516\n",
      "0.0050013065338134766\n",
      "0.004500150680541992\n",
      "0.005000591278076172\n",
      "0.005000591278076172\n",
      "0.0050013065338134766\n",
      "0.004499673843383789\n",
      "0.0045011043548583984\n",
      "0.004500865936279297\n",
      "0.005501508712768555\n",
      "0.005000114440917969\n",
      "0.0045011043548583984\n",
      "0.004500150680541992\n",
      "0.004500627517700195\n",
      "0.005000114440917969\n",
      "0.005001068115234375\n",
      "0.005000591278076172\n",
      "0.005501747131347656\n",
      "0.00500035285949707\n",
      "0.005000591278076172\n",
      "0.005001068115234375\n",
      "0.0050008296966552734\n",
      "0.005000114440917969\n",
      "0.004500627517700195\n",
      "0.005001068115234375\n",
      "0.0055010318756103516\n",
      "0.004500627517700195\n",
      "0.004500627517700195\n",
      "0.00550079345703125\n",
      "0.004000663757324219\n",
      "0.004500389099121094\n",
      "0.0045011043548583984\n",
      "0.005000591278076172\n",
      "0.004500865936279297\n",
      "0.0050008296966552734\n",
      "0.004500627517700195\n",
      "0.04700779914855957\n",
      "0.006001472473144531\n",
      "0.005000591278076172\n",
      "0.005000591278076172\n",
      "0.0050013065338134766\n",
      "0.004500150680541992\n",
      "0.005001068115234375\n",
      "0.004499673843383789\n",
      "0.0050013065338134766\n",
      "0.004999876022338867\n",
      "0.0050013065338134766\n",
      "0.005000591278076172\n",
      "0.0050013065338134766\n",
      "0.004499912261962891\n",
      "0.0050008296966552734\n",
      "0.0050008296966552734\n",
      "0.0050008296966552734\n",
      "0.0050008296966552734\n",
      "0.004999637603759766\n",
      "0.004500865936279297\n",
      "0.004500627517700195\n",
      "0.0050008296966552734\n",
      "0.005501508712768555\n",
      "0.00500035285949707\n",
      "0.004500865936279297\n",
      "0.0050008296966552734\n",
      "0.00500035285949707\n",
      "0.005501747131347656\n",
      "0.004500150680541992\n",
      "0.0050013065338134766\n",
      "0.005501985549926758\n",
      "0.004999876022338867\n",
      "0.005501508712768555\n",
      "0.005000114440917969\n",
      "0.004500150680541992\n",
      "0.0050008296966552734\n",
      "0.005001068115234375\n",
      "0.0050008296966552734\n",
      "0.005001068115234375\n",
      "0.004999876022338867\n",
      "0.0045011043548583984\n",
      "0.004500627517700195\n",
      "0.005001068115234375\n",
      "0.006501197814941406\n",
      "0.005501747131347656\n",
      "0.04650759696960449\n",
      "0.004499912261962891\n",
      "0.004500150680541992\n",
      "0.005000591278076172\n",
      "0.0050013065338134766\n",
      "0.00500035285949707\n",
      "0.004500865936279297\n",
      "0.005000114440917969\n",
      "0.005001068115234375\n",
      "0.004500865936279297\n",
      "0.0050008296966552734\n",
      "0.004500389099121094\n",
      "0.0050008296966552734\n",
      "0.004500627517700195\n",
      "0.0050008296966552734\n",
      "0.004500627517700195\n",
      "0.004500389099121094\n",
      "0.0050008296966552734\n",
      "0.005001068115234375\n",
      "0.004500865936279297\n",
      "0.005000591278076172\n",
      "0.0050008296966552734\n",
      "0.005001068115234375\n",
      "0.0050008296966552734\n",
      "0.004500865936279297\n",
      "0.004500389099121094\n",
      "0.005501508712768555\n",
      "0.00500035285949707\n",
      "0.004499912261962891\n",
      "0.004500627517700195\n",
      "0.004500627517700195\n",
      "0.005000114440917969\n",
      "0.0045011043548583984\n",
      "0.004500150680541992\n",
      "0.004500150680541992\n",
      "0.005000591278076172\n",
      "0.0050008296966552734\n",
      "0.005001068115234375\n",
      "0.0050008296966552734\n",
      "0.005000591278076172\n",
      "0.005501747131347656\n",
      "0.005000114440917969\n",
      "0.0050013065338134766\n",
      "0.0050008296966552734\n",
      "0.00500035285949707\n",
      "0.0045011043548583984\n",
      "0.04700779914855957\n",
      "0.0045011043548583984\n",
      "0.0050013065338134766\n",
      "0.004500627517700195\n",
      "0.0050013065338134766\n",
      "0.00500035285949707\n",
      "0.004500865936279297\n",
      "0.004500627517700195\n",
      "0.0045011043548583984\n",
      "0.004500389099121094\n",
      "0.004500865936279297\n",
      "0.004500627517700195\n",
      "0.0050008296966552734\n",
      "0.005001068115234375\n",
      "0.005000591278076172\n",
      "0.005001068115234375\n",
      "0.004500150680541992\n",
      "0.004500627517700195\n",
      "0.00500035285949707\n",
      "0.004500150680541992\n",
      "0.005001068115234375\n",
      "0.005501747131347656\n",
      "0.004999876022338867\n",
      "0.005001068115234375\n",
      "0.0050008296966552734\n",
      "0.005001068115234375\n",
      "0.005000591278076172\n",
      "0.004499912261962891\n",
      "0.004500627517700195\n",
      "0.004500627517700195\n",
      "0.004500389099121094\n",
      "0.005000591278076172\n",
      "0.0050008296966552734\n",
      "0.004500865936279297\n",
      "0.0050008296966552734\n",
      "0.0050008296966552734\n",
      "0.0050013065338134766\n",
      "0.0050008296966552734\n",
      "0.004999399185180664\n",
      "0.0050008296966552734\n",
      "0.0050008296966552734\n",
      "0.004500389099121094\n",
      "0.0050008296966552734\n",
      "0.004500865936279297\n",
      "0.0050008296966552734\n",
      "0.005000591278076172\n",
      "0.04650759696960449\n",
      "0.0050013065338134766\n",
      "0.005000114440917969\n",
      "0.0045011043548583984\n",
      "0.005000591278076172\n",
      "0.006001472473144531\n",
      "0.005000591278076172\n",
      "0.005001068115234375\n",
      "0.0055010318756103516\n",
      "0.0050008296966552734\n",
      "0.005000591278076172\n",
      "0.005001068115234375\n",
      "0.005001068115234375\n",
      "0.0055005550384521484\n",
      "0.004500627517700195\n",
      "0.005000591278076172\n",
      "0.005001068115234375\n",
      "0.005000591278076172\n",
      "0.005001068115234375\n",
      "0.004500389099121094\n",
      "0.0044994354248046875\n",
      "0.004500389099121094\n",
      "0.0055010318756103516\n",
      "0.004500627517700195\n",
      "0.004999876022338867\n",
      "0.0050008296966552734\n",
      "0.0050013065338134766\n",
      "0.004500389099121094\n",
      "0.005001068115234375\n",
      "0.005000591278076172\n",
      "0.0050013065338134766\n",
      "0.005000591278076172\n",
      "0.004500865936279297\n",
      "0.005000591278076172\n",
      "0.0050008296966552734\n",
      "0.005501508712768555\n",
      "0.005000114440917969\n",
      "0.005001068115234375\n",
      "0.004500627517700195\n",
      "0.0050013065338134766\n",
      "0.00500035285949707\n",
      "0.0050013065338134766\n",
      "0.005000114440917969\n",
      "0.004500865936279297\n",
      "0.004500865936279297\n",
      "0.005501508712768555\n",
      "0.04700803756713867\n",
      "0.0045011043548583984\n",
      "0.004999637603759766\n",
      "0.0045011043548583984\n",
      "0.005000591278076172\n",
      "0.005001068115234375\n",
      "0.0050008296966552734\n",
      "0.0055010318756103516\n",
      "0.004500627517700195\n",
      "0.004500389099121094\n",
      "0.00500035285949707\n",
      "0.0045011043548583984\n",
      "0.004500389099121094\n",
      "0.0045011043548583984\n",
      "0.004500389099121094\n",
      "0.004500865936279297\n",
      "0.005000591278076172\n",
      "0.0050008296966552734\n",
      "0.0050008296966552734\n",
      "0.005001068115234375\n",
      "0.0055010318756103516\n",
      "0.005000591278076172\n",
      "0.005001068115234375\n",
      "0.005001068115234375\n",
      "0.0050008296966552734\n",
      "0.004500389099121094\n",
      "0.0050008296966552734\n",
      "0.0050008296966552734\n",
      "0.005000591278076172\n",
      "0.0050008296966552734\n",
      "0.004500865936279297\n",
      "0.004500627517700195\n",
      "0.004500865936279297\n",
      "0.004500389099121094\n",
      "0.004500150680541992\n",
      "0.005000591278076172\n",
      "0.004999876022338867\n",
      "0.0050008296966552734\n",
      "0.005001068115234375\n",
      "0.005001068115234375\n",
      "0.005001068115234375\n",
      "0.004500150680541992\n",
      "0.004500865936279297\n",
      "0.004500627517700195\n",
      "0.005000591278076172\n",
      "0.0465083122253418\n",
      "0.005001068115234375\n",
      "0.0050008296966552734\n",
      "0.004500865936279297\n",
      "0.004500389099121094\n",
      "0.005001068115234375\n",
      "0.0045013427734375\n",
      "0.00500035285949707\n",
      "0.005000591278076172\n",
      "0.005001068115234375\n",
      "0.004500627517700195\n",
      "0.0050008296966552734\n",
      "0.0050008296966552734\n",
      "0.005501508712768555\n",
      "0.004500150680541992\n",
      "0.004500150680541992\n",
      "0.004500627517700195\n",
      "0.0050008296966552734\n",
      "0.004500865936279297\n",
      "0.0050008296966552734\n",
      "0.005000591278076172\n",
      "0.005501270294189453\n",
      "0.00500035285949707\n",
      "0.0055010318756103516\n",
      "0.00550079345703125\n",
      "0.004500627517700195\n",
      "0.005501508712768555\n",
      "0.00500035285949707\n",
      "0.0050008296966552734\n",
      "0.0050008296966552734\n",
      "0.0050008296966552734\n",
      "0.005501270294189453\n",
      "0.0050008296966552734\n",
      "0.004500865936279297\n",
      "0.005000591278076172\n",
      "0.005501270294189453\n",
      "0.005000591278076172\n",
      "0.005001068115234375\n",
      "0.005000591278076172\n",
      "0.005501508712768555\n",
      "0.005000591278076172\n",
      "0.0050008296966552734\n",
      "0.005000591278076172\n",
      "0.0045011043548583984\n",
      "0.004500389099121094\n",
      "0.004500865936279297\n",
      "0.04850888252258301\n",
      "0.0050008296966552734\n",
      "0.004500865936279297\n",
      "0.00500035285949707\n",
      "0.004500865936279297\n",
      "0.004500389099121094\n",
      "0.0045013427734375\n",
      "0.004999876022338867\n",
      "0.005501270294189453\n",
      "0.0050008296966552734\n",
      "0.004499912261962891\n",
      "0.0050008296966552734\n",
      "0.0050008296966552734\n",
      "0.0050008296966552734\n",
      "0.004500389099121094\n",
      "0.004500627517700195\n",
      "0.004500865936279297\n",
      "0.004500389099121094\n",
      "0.004500150680541992\n",
      "0.005000114440917969\n",
      "0.005501747131347656\n",
      "0.00500035285949707\n",
      "0.004500627517700195\n",
      "0.004500627517700195\n",
      "0.005001068115234375\n",
      "0.0055005550384521484\n",
      "0.005001544952392578\n",
      "0.00500035285949707\n",
      "0.0050008296966552734\n",
      "0.0050008296966552734\n",
      "0.005501508712768555\n",
      "0.005000591278076172\n",
      "0.004500627517700195\n",
      "0.0045013427734375\n",
      "0.005000114440917969\n",
      "0.005001068115234375\n",
      "0.005001068115234375\n",
      "0.0050008296966552734\n",
      "0.0050008296966552734\n",
      "0.0050008296966552734\n",
      "0.00500035285949707\n",
      "0.004500627517700195\n",
      "0.005000591278076172\n",
      "0.0050013065338134766\n",
      "0.004500150680541992\n",
      "0.04750776290893555\n",
      "0.0045011043548583984\n",
      "0.005001544952392578\n",
      "0.0050008296966552734\n",
      "0.004500627517700195\n",
      "0.005001068115234375\n",
      "0.00500178337097168\n",
      "0.004999876022338867\n",
      "0.005001544952392578\n",
      "0.005000114440917969\n",
      "0.00500178337097168\n",
      "0.00500035285949707\n",
      "0.005001068115234375\n",
      "0.0045011043548583984\n",
      "0.0045013427734375\n",
      "0.004500389099121094\n",
      "0.0045013427734375\n",
      "0.004500865936279297\n",
      "0.005001068115234375\n",
      "0.004500150680541992\n",
      "0.005001544952392578\n",
      "0.0055005550384521484\n",
      "0.0050013065338134766\n",
      "0.00550079345703125\n",
      "0.005001068115234375\n",
      "0.0055005550384521484\n",
      "0.004500865936279297\n",
      "0.0050008296966552734\n",
      "0.005000591278076172\n",
      "0.006000995635986328\n",
      "0.0055010318756103516\n",
      "0.0045011043548583984\n",
      "0.0050013065338134766\n",
      "0.004500865936279297\n",
      "0.0050013065338134766\n",
      "0.0045013427734375\n",
      "0.004500627517700195\n",
      "0.0050013065338134766\n",
      "0.0055005550384521484\n",
      "0.004500865936279297\n",
      "0.004500865936279297\n",
      "0.0045011043548583984\n",
      "0.0050008296966552734\n",
      "0.005000591278076172\n",
      "0.004500389099121094\n",
      "0.005001068115234375\n",
      "0.0465085506439209\n",
      "0.004500627517700195\n",
      "0.0055005550384521484\n",
      "0.0045013427734375\n",
      "0.004500865936279297\n",
      "0.004500865936279297\n",
      "0.004500865936279297\n",
      "0.0050013065338134766\n",
      "0.005500316619873047\n",
      "0.0045011043548583984\n",
      "0.004500627517700195\n",
      "0.004500865936279297\n",
      "0.004500389099121094\n",
      "0.005001544952392578\n",
      "0.0060002803802490234\n",
      "0.005001544952392578\n",
      "0.004500865936279297\n",
      "0.0055005550384521484\n",
      "0.005001544952392578\n",
      "0.004500389099121094\n",
      "0.004500865936279297\n",
      "0.0045011043548583984\n",
      "0.0050008296966552734\n",
      "0.004500627517700195\n",
      "0.005000114440917969\n",
      "0.0045011043548583984\n",
      "0.004501819610595703\n",
      "0.00500035285949707\n",
      "0.004499673843383789\n",
      "0.005001068115234375\n",
      "0.004500150680541992\n",
      "0.0050013065338134766\n",
      "0.005000591278076172\n",
      "0.005001544952392578\n",
      "0.005000114440917969\n",
      "0.0045015811920166016\n",
      "0.004500627517700195\n",
      "0.0045011043548583984\n",
      "0.0050013065338134766\n",
      "0.005001068115234375\n",
      "0.0045011043548583984\n",
      "0.004500627517700195\n",
      "0.0050008296966552734\n",
      "0.004500865936279297\n",
      "0.004500389099121094\n",
      "0.04700803756713867\n",
      "0.004500150680541992\n",
      "0.0050013065338134766\n",
      "0.005001544952392578\n",
      "0.005500316619873047\n",
      "0.005001544952392578\n",
      "0.004500627517700195\n",
      "0.0050008296966552734\n",
      "0.004500627517700195\n",
      "0.0050013065338134766\n",
      "0.004500150680541992\n",
      "0.0045013427734375\n",
      "0.0050008296966552734\n",
      "0.0045013427734375\n",
      "0.0055005550384521484\n",
      "0.0045011043548583984\n",
      "0.0050008296966552734\n",
      "0.005001068115234375\n",
      "0.004500865936279297\n",
      "0.004500865936279297\n",
      "0.0050008296966552734\n",
      "0.0050008296966552734\n",
      "0.0045011043548583984\n",
      "0.00550079345703125\n",
      "0.004500627517700195\n",
      "0.005001068115234375\n",
      "0.0050008296966552734\n",
      "0.005001068115234375\n",
      "0.004500627517700195\n",
      "0.0045011043548583984\n",
      "0.0045013427734375\n",
      "0.004500865936279297\n",
      "0.0060007572174072266\n",
      "0.005001068115234375\n",
      "0.004501819610595703\n",
      "0.0045011043548583984\n",
      "0.005000114440917969\n",
      "0.0050008296966552734\n",
      "0.0050008296966552734\n",
      "0.005001544952392578\n",
      "0.004500865936279297\n",
      "0.005001068115234375\n",
      "0.005500316619873047\n",
      "0.006500959396362305\n",
      "0.008501768112182617\n",
      "0.0060007572174072266\n",
      "0.04800891876220703\n",
      "0.004500627517700195\n",
      "0.0045013427734375\n",
      "0.004501819610595703\n",
      "0.004999637603759766\n",
      "0.005001544952392578\n",
      "0.0055010318756103516\n",
      "0.0050008296966552734\n",
      "0.0050013065338134766\n",
      "0.0050008296966552734\n",
      "0.005000591278076172\n",
      "0.0050008296966552734\n",
      "0.0045011043548583984\n",
      "0.005001068115234375\n",
      "0.00500035285949707\n",
      "0.004500389099121094\n",
      "0.004500865936279297\n",
      "0.005002021789550781\n",
      "0.004999876022338867\n",
      "0.0050008296966552734\n",
      "0.0050008296966552734\n",
      "0.005501270294189453\n",
      "0.004500865936279297\n",
      "0.004500627517700195\n",
      "0.005000591278076172\n",
      "0.005001068115234375\n",
      "0.0045015811920166016\n",
      "0.004500389099121094\n",
      "0.00500178337097168\n",
      "0.004500389099121094\n",
      "0.004500865936279297\n",
      "0.005001544952392578\n",
      "0.005000114440917969\n",
      "0.004500627517700195\n",
      "0.005001068115234375\n",
      "0.0050013065338134766\n",
      "0.004500865936279297\n",
      "0.004500865936279297\n",
      "0.004500150680541992\n",
      "0.0045011043548583984\n",
      "0.0055010318756103516\n",
      "0.004500627517700195\n",
      "0.004501819610595703\n",
      "0.005501270294189453\n",
      "0.00500178337097168\n",
      "0.046507835388183594\n",
      "0.004500389099121094\n",
      "0.004500865936279297\n",
      "0.0050008296966552734\n",
      "0.004500865936279297\n",
      "0.005001068115234375\n",
      "0.0050013065338134766\n",
      "0.004500389099121094\n",
      "0.005001068115234375\n",
      "0.0050008296966552734\n",
      "0.005000591278076172\n",
      "0.0045011043548583984\n",
      "0.004500865936279297\n",
      "0.0045013427734375\n",
      "0.005001068115234375\n",
      "0.005001068115234375\n",
      "0.0050008296966552734\n",
      "0.006000041961669922\n",
      "0.0055005550384521484\n",
      "0.0050013065338134766\n",
      "0.0050008296966552734\n",
      "0.0055010318756103516\n",
      "0.005001068115234375\n",
      "0.005000591278076172\n",
      "0.004500150680541992\n",
      "0.0045013427734375\n",
      "0.0045013427734375\n",
      "0.0055010318756103516\n",
      "0.005000591278076172\n",
      "0.00500178337097168\n",
      "0.005000114440917969\n",
      "0.004501819610595703\n",
      "0.0045011043548583984\n",
      "0.004500389099121094\n",
      "0.004500389099121094\n",
      "0.005501985549926758\n",
      "0.005080699920654297\n",
      "0.00492095947265625\n",
      "0.004500389099121094\n",
      "0.0045013427734375\n",
      "0.005001544952392578\n",
      "0.00500035285949707\n",
      "0.0050008296966552734\n",
      "0.004500627517700195\n",
      "0.0050008296966552734\n",
      "0.00500035285949707\n",
      "0.04700827598571777\n",
      "0.0045015811920166016\n",
      "0.0045013427734375\n",
      "0.004499673843383789\n",
      "0.004500865936279297\n",
      "0.006000518798828125\n",
      "0.0050013065338134766\n",
      "0.0050013065338134766\n",
      "0.006000995635986328\n",
      "0.004500627517700195\n",
      "0.004500627517700195\n",
      "0.00500035285949707\n",
      "0.0050008296966552734\n",
      "0.00500035285949707\n",
      "0.005001068115234375\n",
      "0.00500035285949707\n",
      "0.0045011043548583984\n",
      "0.00500178337097168\n",
      "0.00500035285949707\n",
      "0.004000425338745117\n",
      "0.005001068115234375\n",
      "0.004499912261962891\n",
      "0.005001068115234375\n",
      "0.005001068115234375\n",
      "0.005001068115234375\n",
      "0.0055005550384521484\n",
      "0.0050008296966552734\n",
      "0.005001068115234375\n",
      "0.0050013065338134766\n",
      "0.005000591278076172\n",
      "0.004500389099121094\n",
      "0.0045015811920166016\n",
      "0.0045013427734375\n",
      "0.0045013427734375\n",
      "0.004500865936279297\n",
      "0.004499912261962891\n",
      "0.004500627517700195\n",
      "0.0050013065338134766\n",
      "0.005001068115234375\n",
      "0.0045011043548583984\n",
      "0.004500865936279297\n",
      "0.0045013427734375\n",
      "0.005001544952392578\n",
      "0.005000591278076172\n",
      "0.00550079345703125\n",
      "0.005001068115234375\n",
      "0.0465085506439209\n",
      "0.0045013427734375\n",
      "0.004500627517700195\n",
      "0.005001544952392578\n",
      "0.0050008296966552734\n",
      "0.0050013065338134766\n",
      "0.00500035285949707\n",
      "0.0045013427734375\n",
      "0.004500865936279297\n",
      "0.0045011043548583984\n",
      "0.004500627517700195\n",
      "0.004500865936279297\n",
      "0.005501508712768555\n",
      "0.004500150680541992\n",
      "0.00500035285949707\n",
      "0.005001544952392578\n",
      "0.0045013427734375\n",
      "0.004501819610595703\n",
      "0.0055010318756103516\n",
      "0.005000591278076172\n",
      "0.004500389099121094\n",
      "0.004500389099121094\n",
      "0.005000114440917969\n",
      "0.005501747131347656\n",
      "0.005001068115234375\n",
      "0.005000591278076172\n",
      "0.0045011043548583984\n",
      "0.004500865936279297\n",
      "0.0050008296966552734\n",
      "0.005001068115234375\n",
      "0.005000591278076172\n",
      "0.0050008296966552734\n",
      "0.0050013065338134766\n",
      "0.00500178337097168\n",
      "0.005499839782714844\n",
      "0.0045011043548583984\n",
      "0.0045015811920166016\n",
      "0.005501508712768555\n",
      "0.004500389099121094\n",
      "0.0050013065338134766\n",
      "0.00500035285949707\n",
      "0.004500865936279297\n",
      "0.0050008296966552734\n",
      "0.005001068115234375\n",
      "0.004500865936279297\n",
      "0.005001068115234375\n",
      "0.005000114440917969\n",
      "0.04700827598571777\n",
      "0.0045011043548583984\n",
      "0.0050008296966552734\n",
      "0.005001068115234375\n",
      "0.005001068115234375\n",
      "0.005000591278076172\n",
      "0.0050008296966552734\n",
      "0.004500627517700195\n",
      "0.004500389099121094\n",
      "0.0045011043548583984\n",
      "0.0050008296966552734\n",
      "0.006000995635986328\n",
      "0.005001544952392578\n",
      "0.005001068115234375\n",
      "0.005000114440917969\n",
      "0.00550079345703125\n",
      "0.0050008296966552734\n",
      "0.0050008296966552734\n",
      "0.0050013065338134766\n",
      "0.005500316619873047\n",
      "0.0050013065338134766\n",
      "0.005000591278076172\n",
      "0.005001068115234375\n",
      "0.005001068115234375\n",
      "0.005000591278076172\n",
      "0.0050008296966552734\n",
      "0.0045011043548583984\n",
      "0.00500178337097168\n",
      "0.004499673843383789\n",
      "0.0050013065338134766\n",
      "0.0045011043548583984\n",
      "0.005000591278076172\n",
      "0.005000591278076172\n",
      "0.005001068115234375\n",
      "0.0045011043548583984\n",
      "0.0045013427734375\n",
      "0.0055010318756103516\n",
      "0.004500865936279297\n",
      "0.0045015811920166016\n",
      "0.005001544952392578\n",
      "0.0050008296966552734\n",
      "0.004500865936279297\n",
      "0.004500389099121094\n",
      "0.0050013065338134766\n",
      "0.00500035285949707\n",
      "0.047008514404296875\n",
      "0.00500178337097168\n",
      "0.004500389099121094\n",
      "0.005000591278076172\n",
      "0.005501508712768555\n",
      "0.005000114440917969\n",
      "0.005001068115234375\n",
      "0.0055010318756103516\n",
      "0.0045011043548583984\n",
      "0.004500389099121094\n",
      "0.005001544952392578\n",
      "0.004500865936279297\n",
      "0.0045015811920166016\n",
      "0.0045011043548583984\n",
      "0.0050008296966552734\n",
      "0.005001068115234375\n",
      "0.005000591278076172\n",
      "0.004500150680541992\n",
      "0.0050013065338134766\n",
      "0.005500316619873047\n",
      "0.005001544952392578\n",
      "0.005001068115234375\n",
      "0.0050008296966552734\n",
      "0.004500627517700195\n",
      "0.0045015811920166016\n",
      "0.0055010318756103516\n",
      "0.004500865936279297\n",
      "0.004500865936279297\n",
      "0.004500627517700195\n",
      "0.0045011043548583984\n",
      "0.005000591278076172\n",
      "0.0045015811920166016\n",
      "0.005500316619873047\n",
      "0.0050013065338134766\n",
      "0.004500389099121094\n",
      "0.0050008296966552734\n",
      "0.0050008296966552734\n",
      "0.005001544952392578\n",
      "0.004500627517700195\n",
      "0.0045013427734375\n",
      "0.0045015811920166016\n",
      "0.004500865936279297\n",
      "0.005001544952392578\n",
      "0.004500865936279297\n",
      "0.0050013065338134766\n",
      "0.005000591278076172\n",
      "0.04750800132751465\n",
      "0.0045011043548583984\n",
      "0.0050013065338134766\n",
      "0.00500035285949707\n",
      "0.0050013065338134766\n",
      "0.005000591278076172\n",
      "0.00500178337097168\n",
      "0.004500865936279297\n",
      "0.0055010318756103516\n",
      "0.0050008296966552734\n",
      "0.005001544952392578\n",
      "0.005000591278076172\n",
      "0.004500389099121094\n",
      "0.004500865936279297\n",
      "0.00500035285949707\n",
      "0.005001544952392578\n",
      "0.005001068115234375\n",
      "0.0045011043548583984\n",
      "0.004500627517700195\n",
      "0.0050008296966552734\n",
      "0.005001068115234375\n",
      "0.005000591278076172\n",
      "0.004500389099121094\n",
      "0.00500035285949707\n",
      "0.0045015811920166016\n",
      "0.004500627517700195\n",
      "0.0045011043548583984\n",
      "0.0050008296966552734\n",
      "0.0045011043548583984\n",
      "0.00500035285949707\n",
      "0.006001472473144531\n",
      "0.00500035285949707\n",
      "0.0045011043548583984\n",
      "0.005001068115234375\n",
      "0.0055005550384521484\n",
      "0.0045015811920166016\n",
      "0.00500035285949707\n",
      "0.0055010318756103516\n",
      "0.0050008296966552734\n",
      "0.005001068115234375\n",
      "0.004500389099121094\n",
      "0.0045011043548583984\n",
      "0.00500178337097168\n",
      "0.0050008296966552734\n",
      "0.0050008296966552734\n",
      "0.0050008296966552734\n",
      "0.048007965087890625\n",
      "0.0050013065338134766\n",
      "0.0050008296966552734\n",
      "0.004500865936279297\n",
      "0.005001544952392578\n",
      "0.004500389099121094\n",
      "0.004500865936279297\n",
      "0.0050013065338134766\n",
      "0.005000591278076172\n",
      "0.004500627517700195\n",
      "0.004500865936279297\n",
      "0.0050008296966552734\n",
      "0.00500035285949707\n",
      "0.005001068115234375\n",
      "0.0050008296966552734\n",
      "0.0045011043548583984\n",
      "0.004500627517700195\n",
      "0.005001068115234375\n",
      "0.004500865936279297\n",
      "0.0045013427734375\n",
      "0.005000591278076172\n",
      "0.005001068115234375\n",
      "0.0050008296966552734\n",
      "0.0050008296966552734\n",
      "0.004500627517700195\n",
      "0.0045015811920166016\n",
      "0.0045011043548583984\n",
      "0.005001544952392578\n",
      "0.005000591278076172\n",
      "0.004500865936279297\n",
      "0.00500035285949707\n",
      "0.0050013065338134766\n",
      "0.005500078201293945\n",
      "0.0050008296966552734\n",
      "0.004500865936279297\n",
      "0.0050008296966552734\n",
      "0.005001068115234375\n",
      "0.004500389099121094\n",
      "0.005002021789550781\n",
      "0.005000591278076172\n",
      "0.00500178337097168\n",
      "0.005000114440917969\n",
      "0.005001068115234375\n",
      "0.0055005550384521484\n",
      "0.0050013065338134766\n",
      "0.0050008296966552734\n",
      "0.04800820350646973\n",
      "0.0055010318756103516\n",
      "0.0045013427734375\n",
      "0.005001068115234375\n",
      "0.0050008296966552734\n",
      "0.00500178337097168\n",
      "0.0050008296966552734\n",
      "0.0050008296966552734\n",
      "0.005001068115234375\n",
      "0.005000591278076172\n",
      "0.005001068115234375\n",
      "0.0044994354248046875\n",
      "0.00500178337097168\n",
      "0\n",
      "14\n"
     ]
    }
   ],
   "source": [
    "k = 8\n",
    "genome = readFasta(\"C_glabrata_1000bp_only.fasta\")\n",
    "print(len(genome))\n",
    "res = {}\n",
    "cpt = 0\n",
    "num_seq = 0\n",
    "for one_genome in genome:\n",
    "    #print(len(one_genome))\n",
    "    start = time.time()\n",
    "    for motif in liste_res:\n",
    "        pos = searchIndexTable(one_genome, k, motif)\n",
    "        if pos != -1:\n",
    "            if motif not in res:\n",
    "                res[motif] = [num_seq]\n",
    "            else:\n",
    "                res[motif].append(num_seq)\n",
    "    num_seq += 1\n",
    "\n",
    "    end = time.time()\n",
    "    print(end-start)\n",
    "print(cpt)\n",
    "print(len(res))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def readProteinProduct(genome):\n",
    "    sequence = []\n",
    "    file = open(genome, \"r\")\n",
    "    proteinNames = []\n",
    "    for s in file:\n",
    "        if s[0] == \">\":\n",
    "            proteinNames.append(s.strip().upper()[1:])\n",
    "    return proteinNames\n",
    "\n",
    "proteinProduct = readProteinProduct(\"C_glabrata_1000bp_only.fasta\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "164\n['CAGL0L04202G', 'CAGL0D04576G', 'CAGL0L07546G', 'CAGL0G07623G', 'CAGL0B00506G', 'CAGL0G02299G', 'CAGL0C04631G', 'CAGL0C00407G', 'CAGL0J01001G', 'CAGL0H06831G', 'CAGL0C00561G', 'CAGL0B04411G', 'CAGL0C01067G', 'CAGL0C00627G', 'CAGL0G05005G', 'CAGL0K03905G', 'CAGL0K11748G', 'CAGL0A01826G', 'CAGL0K06149G', 'CAGL0I10560G', 'CAGL0D04444G', 'CAGL0M12947G', 'CAGL0H05005G', 'CAGL0L09647G', 'CAGL0F05027G', 'CAGL0G04191G', 'CAGL0L00869G', 'CAGL0I10967G', 'CAGL0B04565G', 'CAGL0H10538G', 'CAGL0L03784G', 'CAGL0F03597G', 'CAGL0G09262G', 'CAGL0M12001G', 'CAGL0E05258G', 'CAGL0I10450G', 'CAGL0H02827G', 'CAGL0J02398G', 'CAGL0J03806G', 'CAGL0M12683G', 'CAGL0J05500G', 'CAGL0C04719G', 'CAGL0H10340G', 'CAGL0J04004G', 'CAGL0B01144G', 'CAGL0I06743G', 'CAGL0L01397G', 'CAGL0H00572G', 'CAGL0K06303G', 'CAGL0J06050G', 'CAGL0F07755G', 'CAGL0G05115G', 'CAGL0F04411G', 'CAGL0L01529G', 'CAGL0K05753G', 'CAGL0D02244G', 'CAGL0H02937G', 'CAGL0I02750G', 'CAGL0C01727G', 'CAGL0H09680G', 'CAGL0K08316G', 'CAGL0G09273G', 'CAGL0A01133G', 'CAGL0G04631G', 'CAGL0L03047G', 'CAGL0L10318G', 'CAGL0K09526G', 'CAGL0I04422G', 'CAGL0K12474G', 'CAGL0L01683G', 'CAGL0E01991G', 'CAGL0D02728G', 'CAGL0I05324G', 'CAGL0D01100G', 'CAGL0J09988G', 'CAGL0I02596G', 'CAGL0L01969G', 'CAGL0E04334G', 'CAGL0J02970G', 'CAGL0L12320G', 'CAGL0F07579G', 'CAGL0I04092G', 'CAGL0C01056G', 'CAGL0B03949G', 'CAGL0L07436G', 'CAGL0F06369G', 'CAGL0K05863G', 'CAGL0F06875G', 'CAGL0J04048G', 'CAGL0J08459G', 'CAGL0A03883G', 'CAGL0I05478G', 'CAGL0J06270G', 'CAGL0J11330G', 'CAGL0D03630G', 'CAGL0D04708G', 'CAGL0L04334G', 'CAGL0C02563G', 'CAGL0J09482G', 'CAGL0I06380G', 'CAGL0F00979G', 'CAGL0M13849G', 'CAGL0B04851G', 'CAGL0F00209G', 'CAGL0G10021G', 'CAGL0L09735G', 'CAGL0M12771G', 'CAGL0F07645G', 'CAGL0I00506G', 'CAGL0L05984G', 'CAGL0M03971G', 'CAGL0C01353G', 'CAGL0A03784G', 'CAGL0K00517G', 'CAGL0I04378G', 'CAGL0D00704G', 'CAGL0L08866G', 'CAGL0E01859G', 'CAGL0E02343G', 'CAGL0J09064G', 'CAGL0J09372G', 'CAGL0E01441G', 'CAGL0F05643G', 'CAGL0K05511G', 'CAGL0D01672G', 'CAGL0I02464G', 'CAGL0M04037G', 'CAGL0M03245G', 'CAGL0K06391G', 'CAGL0D01364G', 'CAGL0L06710G', 'CAGL0J08657G', 'CAGL0B02211G', 'CAGL0E03498G', 'CAGL0I09086G', 'CAGL0K02013G', 'CAGL0J02310G', 'CAGL0I04444G', 'CAGL0B02167G', 'CAGL0L12474G', 'CAGL0F04873G', 'CAGL0K03729G', 'CAGL0K04697G', 'CAGL0H10202G', 'CAGL0B04939G', 'CAGL0L09845G', 'CAGL0G03245G', 'CAGL0D03190G', 'CAGL0M01518G', 'CAGL0M12991G', 'CAGL0K13024G', 'CAGL0I06644G', 'CAGL0J00968G', 'CAGL0I07623G', 'CAGL0M04213G', 'CAGL0L08888G', 'CAGL0H07953G', 'CAGL0K09614G', 'CAGL0F03443G', 'CAGL0M10527G', 'CAGL0C01133G', 'CAGL0F02805G', 'CAGL0I05258G', 'CAGL0H03861G']\n"
     ]
    }
   ],
   "source": [
    "liste_protein_product_mac1 = []\n",
    "for num_sequence in res[motif_mac1]:\n",
    "    liste_protein_product_mac1.append(proteinProduct[num_sequence])\n",
    "\n",
    "print(len(liste_protein_product_mac1))\n",
    "print(liste_protein_product_mac1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "159\n['UNCHARACTERIZED, PROTEIN, CAGL0K13024G', 'UNCHARACTERIZED, PROTEIN, CAGL0A01133G', 'UNCHARACTERIZED, PROTEIN, CAGL0A01826G', 'UNCHARACTERIZED, PROTEIN, CAGL0I00506G', 'UNCHARACTERIZED, PROTEIN, CAGL0I02464G', 'UNCHARACTERIZED, PROTEIN, CAGL0I02596G', 'UNCHARACTERIZED, PROTEIN, CAGL0I02750G', 'UNCHARACTERIZED, PROTEIN, CAGL0I04092G', 'UNCHARACTERIZED, PROTEIN, CAGL0I04378G', 'UNCHARACTERIZED, PROTEIN, CAGL0I04422G', 'UNCHARACTERIZED, PROTEIN, CAGL0I04444G', 'UNCHARACTERIZED, PROTEIN, CAGL0B00506G', 'UNCHARACTERIZED, PROTEIN, CAGL0B01144G', 'UNCHARACTERIZED, PROTEIN, CAGL0I05258G', 'UNCHARACTERIZED, PROTEIN, CAGL0I05324G', 'UNCHARACTERIZED, PROTEIN, CAGL0I05478G', 'MITOCHONDRIAL, 54S, RIBOSOMAL, PROTEIN, YML9', 'UNCHARACTERIZED, PROTEIN, CAGL0B02211G', 'UNCHARACTERIZED, PROTEIN, CAGL0F00209G', 'UNCHARACTERIZED, PROTEIN, CAGL0F00979G', 'UNCHARACTERIZED, PROTEIN, CAGL0I06380G', 'UNCHARACTERIZED, PROTEIN, CAGL0I06644G', 'UNCHARACTERIZED, PROTEIN, CAGL0I06743G', 'UNCHARACTERIZED, PROTEIN, CAGL0I07623G', 'UNCHARACTERIZED, PROTEIN, CAGL0B03949G', 'UNCHARACTERIZED, PROTEIN, CAGL0B04411G', 'UNCHARACTERIZED, PROTEIN, CAGL0B04565G', 'UNCHARACTERIZED, PROTEIN, CAGL0A03784G', 'UNCHARACTERIZED, PROTEIN, CAGL0A03883G', 'UNCHARACTERIZED, PROTEIN, CAGL0I09086G', 'UNCHARACTERIZED, PROTEIN, CAGL0F02805G', 'UNCHARACTERIZED, PROTEIN, CAGL0I10450G', 'UNCHARACTERIZED, PROTEIN, CAGL0I10560G', 'UNCHARACTERIZED, PROTEIN, CAGL0G02299G', 'UNCHARACTERIZED, PROTEIN, CAGL0G03245G', 'UNCHARACTERIZED, PROTEIN, CAGL0F03443G', 'UNCHARACTERIZED, PROTEIN, CAGL0F03597G', 'MITOCHONDRIAL, 54S, RIBOSOMAL, PROTEIN, YML47', 'UNCHARACTERIZED, PROTEIN, CAGL0F04873G', 'UNCHARACTERIZED, PROTEIN, CAGL0F05027G', 'UNCHARACTERIZED, PROTEIN, CAGL0F05643G', 'UNCHARACTERIZED, PROTEIN, CAGL0L00869G', 'UNCHARACTERIZED, PROTEIN, CAGL0L01397G', 'UNCHARACTERIZED, PROTEIN, CAGL0L01529G', 'UNCHARACTERIZED, PROTEIN, CAGL0L01683G', 'UNCHARACTERIZED, PROTEIN, CAGL0L01969G', 'UNCHARACTERIZED, PROTEIN, CAGL0F06369G', 'UNCHARACTERIZED, PROTEIN, CAGL0G04191G', 'UNCHARACTERIZED, PROTEIN, CAGL0G04631G', 'UNCHARACTERIZED, PROTEIN, CAGL0G05005G', 'UNCHARACTERIZED, PROTEIN, CAGL0G05115G', 'UNCHARACTERIZED, PROTEIN, CAGL0L03047G', 'UNCHARACTERIZED, PROTEIN, CAGL0L03784G', 'UNCHARACTERIZED, PROTEIN, CAGL0L04202G', 'MITOCHONDRIAL, 37S, RIBOSOMAL, PROTEIN, PET123', 'UNCHARACTERIZED, PROTEIN, CAGL0F06875G', 'UNCHARACTERIZED, PROTEIN, CAGL0F07579G', 'UNCHARACTERIZED, PROTEIN, CAGL0F07645G', 'UNCHARACTERIZED, PROTEIN, CAGL0F07755G', 'UNCHARACTERIZED, PROTEIN, CAGL0G07623G', 'UNCHARACTERIZED, PROTEIN, CAGL0G09273G', 'UNCHARACTERIZED, PROTEIN, CAGL0L05984G', 'UNCHARACTERIZED, PROTEIN, CAGL0L06710G', 'UNCHARACTERIZED, PROTEIN, CAGL0L07436G', 'UNCHARACTERIZED, PROTEIN, CAGL0L07546G', 'UNCHARACTERIZED, PROTEIN, CAGL0J01001G', 'UNCHARACTERIZED, PROTEIN, CAGL0L08866G', 'UNCHARACTERIZED, PROTEIN, CAGL0L08888G', 'UNCHARACTERIZED, PROTEIN, CAGL0L09647G', 'UNCHARACTERIZED, PROTEIN, CAGL0L09735G', 'UNCHARACTERIZED, PROTEIN, CAGL0L09845G', 'UNCHARACTERIZED, PROTEIN, CAGL0L10318G', 'UNCHARACTERIZED, PROTEIN, CAGL0L12320G', 'UNCHARACTERIZED, PROTEIN, CAGL0L12474G', 'UNCHARACTERIZED, PROTEIN, CAGL0J02310G', 'UNCHARACTERIZED, PROTEIN, CAGL0J02398G', 'UNCHARACTERIZED, PROTEIN, CAGL0J02970G', 'UNCHARACTERIZED, PROTEIN, CAGL0J03806G', 'UNCHARACTERIZED, PROTEIN, CAGL0J04004G', 'UNCHARACTERIZED, PROTEIN, CAGL0J04048G', 'UNCHARACTERIZED, PROTEIN, CAGL0J05500G', 'UNCHARACTERIZED, PROTEIN, CAGL0J06050G', 'UNCHARACTERIZED, PROTEIN, CAGL0J06270G', 'UNCHARACTERIZED, PROTEIN, CAGL0E01441G', 'UNCHARACTERIZED, PROTEIN, CAGL0J08459G', 'UNCHARACTERIZED, PROTEIN, CAGL0J08657G', 'UNCHARACTERIZED, PROTEIN, CAGL0J09064G', 'UNCHARACTERIZED, PROTEIN, CAGL0E01859G', '40S, RIBOSOMAL, PROTEIN, S19', 'UNCHARACTERIZED, PROTEIN, CAGL0E02343G', 'UNCHARACTERIZED, PROTEIN, CAGL0J09372G', 'UNCHARACTERIZED, PROTEIN, CAGL0J09482G', 'UNCHARACTERIZED, PROTEIN, CAGL0E03498G', 'UNCHARACTERIZED, PROTEIN, CAGL0J09988G', 'UNCHARACTERIZED, PROTEIN, CAGL0E04334G', 'MITOCHONDRIAL, 54S, RIBOSOMAL, PROTEIN, YML19', 'UNCHARACTERIZED, PROTEIN, CAGL0E05258G', 'UNCHARACTERIZED, PROTEIN, CAGL0I10967G', 'UNCHARACTERIZED, PROTEIN, CAGL0M01518G', 'UNCHARACTERIZED, PROTEIN, CAGL0M03245G', 'UNCHARACTERIZED, PROTEIN, CAGL0M03971G', 'UNCHARACTERIZED, PROTEIN, CAGL0M04037G', 'UNCHARACTERIZED, PROTEIN, CAGL0M04213G', 'UNCHARACTERIZED, PROTEIN, CAGL0M10527G', 'UNCHARACTERIZED, PROTEIN, CAGL0G10021G', 'UNCHARACTERIZED, PROTEIN, CAGL0M12683G', 'UNCHARACTERIZED, PROTEIN, CAGL0M12771G', 'UNCHARACTERIZED, PROTEIN, CAGL0M12947G', 'UNCHARACTERIZED, PROTEIN, CAGL0M12991G', 'UNCHARACTERIZED, PROTEIN, CAGL0K00517G', 'UNCHARACTERIZED, PROTEIN, CAGL0K02013G', 'UNCHARACTERIZED, PROTEIN, CAGL0K03729G', 'UNCHARACTERIZED, PROTEIN, CAGL0K03905G', 'UNCHARACTERIZED, PROTEIN, CAGL0K04697G', 'UNCHARACTERIZED, PROTEIN, CAGL0K05511G', 'UNCHARACTERIZED, PROTEIN, CAGL0K05753G', 'UNCHARACTERIZED, PROTEIN, CAGL0K05863G', '40S, RIBOSOMAL, PROTEIN, S17', 'UNCHARACTERIZED, PROTEIN, CAGL0K06303G', 'MITOCHONDRIAL, 54S, RIBOSOMAL, PROTEIN, YML28', 'UNCHARACTERIZED, PROTEIN, CAGL0K08316G', 'UNCHARACTERIZED, PROTEIN, CAGL0K09526G', 'UNCHARACTERIZED, PROTEIN, CAGL0K09614G', '40S, RIBOSOMAL, PROTEIN, S11', 'UNCHARACTERIZED, PROTEIN, CAGL0H00572G', 'UNCHARACTERIZED, PROTEIN, CAGL0H02827G', 'UNCHARACTERIZED, PROTEIN, CAGL0H02937G', 'UNCHARACTERIZED, PROTEIN, CAGL0H03861G', 'UNCHARACTERIZED, PROTEIN, CAGL0H05005G', 'UNCHARACTERIZED, PROTEIN, CAGL0H06831G', 'UNCHARACTERIZED, PROTEIN, CAGL0H07953G', 'UNCHARACTERIZED, PROTEIN, CAGL0K12474G', 'UNCHARACTERIZED, PROTEIN, CAGL0C00407G', 'UNCHARACTERIZED, PROTEIN, CAGL0C00561G', 'UNCHARACTERIZED, PROTEIN, CAGL0C00627G', 'UNCHARACTERIZED, PROTEIN, CAGL0C01133G', 'UNCHARACTERIZED, PROTEIN, CAGL0C01353G', 'UNCHARACTERIZED, PROTEIN, CAGL0C01727G', 'UNCHARACTERIZED, PROTEIN, CAGL0C02563G', 'UNCHARACTERIZED, PROTEIN, CAGL0H09680G', 'UNCHARACTERIZED, PROTEIN, CAGL0H10202G', 'UNCHARACTERIZED, PROTEIN, CAGL0C04631G', 'UNCHARACTERIZED, PROTEIN, CAGL0C04719G', 'UNCHARACTERIZED, PROTEIN, CAGL0D00704G', 'UNCHARACTERIZED, PROTEIN, CAGL0D01100G', 'UNCHARACTERIZED, PROTEIN, CAGL0D01364G', 'UNCHARACTERIZED, PROTEIN, CAGL0D01672G', 'UNCHARACTERIZED, PROTEIN, CAGL0D02244G', 'UNCHARACTERIZED, PROTEIN, CAGL0D02728G', 'UNCHARACTERIZED, PROTEIN, CAGL0D03190G', 'UNCHARACTERIZED, PROTEIN, CAGL0D03630G', 'UNCHARACTERIZED, PROTEIN, CAGL0D04444G', 'UNCHARACTERIZED, PROTEIN, CAGL0D04576G', 'UNCHARACTERIZED, PROTEIN, CAGL0D04708G', 'UNCHARACTERIZED, PROTEIN, CAGL0M13849G', 'UNCHARACTERIZED, PROTEIN, CAGL0H10340G', 'UNCHARACTERIZED, PROTEIN, CAGL0H10538G', 'UNCHARACTERIZED, PROTEIN, CAGL0B04851G', 'UNCHARACTERIZED, PROTEIN, CAGL0B04939G']\n"
     ]
    }
   ],
   "source": [
    "import csv\n",
    "csv_bien_foutu = []\n",
    "proteinNames = []\n",
    "\n",
    "with open('C_glabrata_annotations.csv', newline='') as csvfile:\n",
    "    annot = csv.reader(csvfile, delimiter=' ', quotechar='|')\n",
    "    for row in annot:\n",
    "        ligne = ', '.join(row).upper()\n",
    "        csv_bien_foutu.append(ligne.split('\"'))\n",
    "for csv_ligne in csv_bien_foutu[1:]:\n",
    "    if csv_ligne[7] in liste_protein_names_mac1:\n",
    "        proteinNames.append(csv_ligne[11])\n",
    "print(len(proteinNames))\n",
    "print(proteinNames)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "celltoolbar": "Raw Cell Format",
  "kernelspec": {
   "name": "python394jvsc74a57bd03729a6ff7676ad95e807d2f5f3db9a400fbe0daa61d45f8837682fe7ac233198",
   "display_name": "Python 3.9.4 64-bit"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.4"
  },
  "metadata": {
   "interpreter": {
    "hash": "3729a6ff7676ad95e807d2f5f3db9a400fbe0daa61d45f8837682fe7ac233198"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}